[4] Re-Implement Empty State for Post Installation Messages screen

PhilETaylor opened the pull request:
Closes  #34048

### Summary of Changes

Correctly Implement Empty State for Post Installation Messages. 

Previously I implemented this before Empty State design was cemented as a part of Joomla 4, so its not using the same layouts as all the other extensions, this PR makes it now consistent with them.

### Testing Instructions

Hide all your post installation messages

### Actual result BEFORE applying this Pull Request

<img width="1603" alt="Screenshot 2021-05-21 at 11 58 55" src="https://user-images.githubusercontent.com/400092/119127212-f0b9ca80-ba2b-11eb-90ad-a485db7acc55.png">


### Expected result AFTER applying this Pull Request

<img width="1602" alt="Screenshot 2021-05-21 at 11 58 35" src="https://user-images.githubusercontent.com/400092/119127179-e4ce0880-ba2b-11eb-88a3-e529fe7ca9f0.png">


### Documentation Changes Required

If you can find a decent documentation page on Post Installation Messages in the doc wiki then let me know and I'll add the Learn More button to it

ChristineWk commented:
I have tested this item :white_check_mark: successfully on e839d4a923ff6a39552476a81db999997317bc52<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

YatharthVyas commented:
I have tested this item :red_circle: unsuccessfully on e839d4a923ff6a39552476a81db999997317bc52<br /><br />When there are 2 or more components having Post Install Messages and if you hide all messages for the 2nd component, you are no longer able to reset them back. Also, If you hide messages for the first Component instead of the second, ie. Hide for Joomla CMS then you are no longer able to navigate to the 2nd Component's Messages.

I'll upload the demonstration of this bug as a gif/image in the next message -><hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

PhilETaylor commented:
ah yeah... forgot about that - I'll mark the PR as draft and handle that - thanks.

YatharthVyas commented:
## Before
![postinstallempty](https://user-images.githubusercontent.com/53610833/119134789-3641bc00-ba5b-11eb-8d7d-adc944170f3e.gif)


## After
![postinstallempty2](https://user-images.githubusercontent.com/53610833/119134777-32159e80-ba5b-11eb-90d4-a7998915ef02.gif)

PhilETaylor commented:
@YatharthVyas can you remember/point me back to my old place I gave you code to add the additional message to your db - save me looking - thanks.

YatharthVyas commented:
@PhilETaylor 

Run this SQL to add a fake post installation message (change the db prefix!)

```
INSERT INTO `o07n8_postinstall_messages` (`postinstall_message_id`, `extension_id`, `title_key`, `description_key`, `action_key`, `language_extension`, `language_client_id`, `type`, `action_file`, `action`, `condition_file`, `condition_method`, `version_introduced`, `enabled`)
VALUES
	(99, 210, 'TEST', 'TEST', '', 'com_cpanel', 1, 'message', '', '', '', '', '4.0.0', 1);
```

Commit:
rework

Signed-off-by: Phil E. Taylor <phil@phil-taylor.com>

PhilETaylor commented:
ok I have reworked this now, it is slight different to the other empty states because the model doesnt extend ListModel for some reason, and even if it did, the getItems doesnt use a listQuery, and even if that did, there is still post-processing of the db results before you finally get a list of messages to show. 

Essentially all this PR is now doing is replacing the hardcoded HTML with the layout file that already exists for empty state

PhilETaylor commented:
When you have more than one extension, of course the whole empty state screen will be lower to give room for the dropdown.

<img width="1597" alt="Screenshot 2021-05-21 at 14 13 27" src="https://user-images.githubusercontent.com/400092/119142770-c58ca680-ba3e-11eb-8ff0-cf96e291295b.png">

YatharthVyas commented:
I have tested this item :white_check_mark: successfully on 4dea2e593ff67f706f50c197125e8d5cbf19c25b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

wilsonge commented on administrator/components/com_postinstall/tmpl/messages/default.php:
```diff
@@ -26,16 +26,11 @@
-		</div>
-	</div>
+	<?php
+	require 'emptystate.php';
```
Not sure if this is consistent with other places but we really should use `HtmlView::loadTemplate()` here so templates can override the subfile without the parent file

PhilETaylor replied:
Cannot use loadTemplate because that will then expect a `default_emptystate.php` file. I'll move it back to layouts like all the other extensions and just move the dropdown into the emptystate.php file.

Quy commented on administrator/components/com_postinstall/tmpl/messages/emptystate.php:
```diff
@@ -0,0 +1,26 @@
+	'icon'       => 'icon-bell',
+];
+
+if (Factory::getApplication()->getIdentity()->authorise('core.admin', 'com_postinstall'))
```
Please check as an administrator can hide all messages, but not reset messages with this PR.

PhilETaylor replied:
it looks like before this there was no ACL checks in the controller at all - are you saying you want an administrator to be able to hide AND reset?

Quy replied:
I’m not sure which is preferred. If can hide, then I assume can reset too. Maybe check using the edit state permission.

PhilETaylor replied:
There was no ACL check previously, so I'll remove my ACL check :) job done:)

sandramay0905 commented:
I have tested this item :white_check_mark: successfully on 4dea2e593ff67f706f50c197125e8d5cbf19c25b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

Commit:
rework the code a different way

PhilETaylor commented:
reworked, to use emptystate layouts like other extensions, only difference is that this needs the additional dropdown if there are multiple extensions with messages

Removed my own ACL check, there was no acl check previously so no need to add one now

Quy commented:
I have tested this item :white_check_mark: successfully on b55056f78050af167b916ad95b3762280489a481<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

jwaisner commented:
I have tested this item :white_check_mark: successfully on b55056f78050af167b916ad95b3762280489a481<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

jwaisner commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34070">issues.joomla.org/tracker/joomla-cms/34070</a>.</sub>

wilsonge commented:
Thanks!
